x/gamm/pool-models/internal/cfmm_common/lp_test.go

```diff
@@ -0,0 +1,117 @@
+package cfmm_common_test
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/osmosis-labs/osmosis/v7/x/gamm/pool-models/balancer"
+	"github.com/osmosis-labs/osmosis/v7/x/gamm/pool-models/internal/cfmm_common"
+	"github.com/osmosis-labs/osmosis/v7/x/gamm/pool-models/stableswap"
+	gammtypes "github.com/osmosis-labs/osmosis/v7/x/gamm/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+func TestCalcExitPool(t *testing.T) {
+	var (
+		two_asset_pool = &stableswap.Pool{
+			TotalShares:   sdk.NewCoin(gammtypes.GetPoolShareDenom(1), gammtypes.InitPoolSharesSupply),
+			PoolLiquidity: sdk.NewCoins(sdk.NewInt64Coin("foo", 1000000000), sdk.NewInt64Coin("bar", 1000000000)),
+		}
+
+		three_asset_pool = &balancer.Pool{
+			TotalShares: sdk.NewCoin(gammtypes.GetPoolShareDenom(1), sdk.NewIntFromUint64(6000000000)),
+			PoolAssets: []balancer.PoolAsset{
+				{Token: sdk.NewInt64Coin("foo", 2000000000)},
+				{Token: sdk.NewInt64Coin("bar", 3000000000)},
+				{Token: sdk.NewInt64Coin("baz", 4000000000)},
+			},
+		}
+	)
+
+	tests := []struct {
+		name          string
+		pool          gammtypes.PoolI
+		exitingShares sdk.Int
+		exitFee       sdk.Dec
+		expExitCoins  sdk.Coins
+		expError      bool
+	}{
+		{
+			name:          "two-asset pool, zero exiting shares",
+			pool:          two_asset_pool,
+			exitingShares: sdk.NewIntFromUint64(0),
+			exitFee:       sdk.ZeroDec(),
+			expExitCoins:  sdk.Coins{},
+			expError:      false,
+		},
+		{
+			name:          "three-asset pool, zero exiting shares",
+			pool:          three_asset_pool,
+			exitingShares: sdk.NewIntFromUint64(0),
+			exitFee:       sdk.ZeroDec(),
+			expExitCoins:  sdk.Coins{},
+			expError:      false,
+		},
+		{
+			name:          "two-asset pool, exiting shares grater than total shares",
+			pool:          two_asset_pool,
+			exitingShares: gammtypes.InitPoolSharesSupply.AddRaw(1),
+			exitFee:       sdk.ZeroDec(),
+			expExitCoins:  sdk.Coins{},
+			expError:      true,
+		},
+		{
+			name:          "three-asset pool, exiting shares grater than total shares",
+			pool:          three_asset_pool,
+			exitingShares: sdk.NewIntFromUint64(7000000000),
+			exitFee:       sdk.ZeroDec(),
+			expExitCoins:  sdk.Coins{},
+			expError:      true,
+		},
+		{
+			name:          "two-asset pool, valid exiting shares",
+			pool:          two_asset_pool,
+			exitingShares: gammtypes.InitPoolSharesSupply.QuoRaw(2),
+			exitFee:       sdk.ZeroDec(),
+			expExitCoins:  sdk.Coins{sdk.NewInt64Coin("foo", 500000000), sdk.NewInt64Coin("bar", 500000000)},
```

Comment:
How were these calculated / estimated?

Is it possible to calculate these based on the initial values when we initialize the pools? That would make it clearer how we got these values and why they are expected. Otherwise, one only can guess where they are coming from

Reply:
> How were these calculated / estimated?

This is calculated simply by hand.
 
> Is it possible to calculate these based on the initial values when we initialize the pools?

U mean defining a function to calculate expected values. I don't know how to elegantly do so without that function looking similar to the `CalcExitPool` function that we're testing.

Reply:
I think I got your idea now

Reply:
Thank you, looks great